sys/xtimer/xtimer.c

```diff
@@ -100,76 +100,43 @@ void _xtimer_periodic_wakeup(uint32_t *last_wakeup, uint32_t period) {
     *last_wakeup = now + offset;
 }
 
-#ifdef MODULE_CORE_MSG
-static void _callback_msg(void* arg)
+#if defined(MODULE_CORE_MSG) && defined(MODULE_CORE_THREAD_FLAGS)
+/* Waits for incoming message or timeout flag. */
+static int _msg_wait(msg_t *m, xtimer_t *t)
 {
-    msg_t *msg = (msg_t*)arg;
-    msg_send_int(msg, msg->sender_pid);
-}
-
-static inline void _setup_msg(xtimer_t *timer, msg_t *msg, kernel_pid_t target_pid)
-{
-    timer->callback = _callback_msg;
-    timer->arg = (void*) msg;
+    int res;
 
-    /* use sender_pid field to get target_pid into callback function */
-    msg->sender_pid = target_pid;
-}
+    thread_flags_t flags =
+        thread_flags_wait_any(THREAD_FLAG_TIMEOUT | THREAD_FLAG_MSG_WAITING);
 
-void _xtimer_set_msg(xtimer_t *timer, uint32_t offset, msg_t *msg, kernel_pid_t target_pid)
-{
-    _setup_msg(timer, msg, target_pid);
-    _xtimer_set64(timer, offset, 0);
-}
-
-void _xtimer_set_msg64(xtimer_t *timer, uint64_t offset, msg_t *msg, kernel_pid_t target_pid)
-{
-    _setup_msg(timer, msg, target_pid);
-    _xtimer_set64(timer, offset, offset >> 32);
-}
-
-/* Prepares the message to trigger the timeout.
- * Additionally, the xtimer_t struct gets initialized.
- */
-static void _setup_timer_msg(msg_t *m, xtimer_t *t)
-{
-    m->type = MSG_XTIMER;
-    m->content.ptr = m;
-
-    t->offset = t->long_offset = 0;
-}
-
-/* Waits for incoming message or timeout. */
-static int _msg_wait(msg_t *m, msg_t *tmsg, xtimer_t *t)
-{
-    msg_receive(m);
-    if (m->type == MSG_XTIMER && m->content.ptr == tmsg) {
-        /* we hit the timeout */
-        return -1;
+    if (flags & THREAD_FLAG_MSG_WAITING) {
+        res = 0;
+        msg_receive(m);
     }
     else {
+        res = -1;
+    }
+
+    if (! (flags & THREAD_FLAG_TIMEOUT)) {
         xtimer_remove(t);
```

Comment:
Hmm.. the timer could fire between the check and calling the remove, leaving a TIMEOUT flag set.

Reply:
Not sure that matters, as the "timeout set" functions always clear that flag at the beginning. I'll clear it anyways.